compiler/rustc_trait_selection/src/traits/misc.rs

```diff
@@ -131,3 +133,19 @@ pub fn type_allowed_to_implement_copy<'tcx>(
 
     Ok(())
 }
+
+pub fn check_const_param_definitely_unequal<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    canonical: Canonical<'tcx, (ParamEnv<'tcx>, Ty<'tcx>, Ty<'tcx>)>,
+) -> Result<(), ()> {
+    let (infcx, (param_env, ty_a, ty_b), _) =
+        tcx.infer_ctxt().build_with_canonical(DUMMY_SP, &canonical);
+    let ocx = ObligationCtxt::new(&infcx);
+
+    let result = ocx.eq(&ObligationCause::dummy(), param_env, ty_a, ty_b);
+    // use `select_where_possible` instead of `select_all_or_error` so that
+    // we don't get errors from obligations being ambiguous.
+    let errors = ocx.select_where_possible();
+
+    if errors.len() > 0 || result.is_err() { Err(()) } else { Ok(()) }
```

Comment:
```suggestion
) -> Result<(), NoSolution> {
    let (infcx, (param_env, ty_a, ty_b), _) =
        tcx.infer_ctxt().build_with_canonical(DUMMY_SP, &canonical);
    let ocx = ObligationCtxt::new(&infcx);

    let result = ocx.eq(&ObligationCause::dummy(), param_env, ty_a, ty_b)?;
    // use `select_where_possible` instead of `select_all_or_error` so that
    // we don't get errors from obligations being ambiguous.
    let errors = ocx.select_where_possible();

    if errors.is_empty() {
        Ok(())
    } else {
        Err(NoSolution)
    }
```

tiny tweak, using `NoSolution` make it a bit cleaner